libpages/response_writer_wrappers.go

```diff
@@ -0,0 +1,44 @@
+package libpages
+
+import "net/http"
+
+var _ http.ResponseWriter = responseWriterWrapper{}
+
+type writeOverride interface {
+	Bytes(req *http.Request) []byte
+}
+
+// responseWriterWrapper wraps a http.ResponseWriter, allowing overriding
+// response for selected status codes. This is mainly for custom error pages.
+//
+// This type is not goroutine safe.
+type responseWriterWrapper struct {
+	w              http.ResponseWriter
+	writeOverrides map[int]writeOverride
+	request        *http.Request
+
+	writtenCode int
+}
+
+// Header implements the http.ResponseWriter interface.
+func (w responseWriterWrapper) Header() http.Header {
+	return w.w.Header()
+}
+
+// WriteHeader implements the http.ResponseWriter interface.
+func (w responseWriterWrapper) WriteHeader(code int) {
+	w.w.WriteHeader(code)
+	w.writtenCode = code
+}
+
+// Write implements the http.ResponseWriter interface.
+func (w responseWriterWrapper) Write(data []byte) (int, error) {
+	if w.writeOverrides == nil || w.writtenCode == 0 {
+		return w.w.Write(data)
+	}
+	override := w.writeOverrides[w.writtenCode]
+	if override == nil {
+		return w.w.Write(data)
```

Comment:
That's what I was saying in standup today, that in non-overridden cases there won't be any additional writes--it's just a passthrough.

Reply:
Yup; as I said in kbchat, I realized later that `WriteHeader` can only happen before `Write` is called, so my concern beforehand was a pointless :)